### PR TITLE
[manager/data_storage] add VineyardBackend and V6D ReportEvent support

### PR DESCRIPTION
**Pull request by lucky-zzz**
- Add VineyardBackend with StandardUri-based vineyard addressing
- Implement CacheManager::ReportEvent (NODE_REGISTER, HEARTBEAT, BLOCK_ADD/DELETE, HOST_DOWN)
- Add CleanupHostLocations for async V6D node removal
- Add MetaSearcher BatchUpsertLocations/BatchDeleteLocations/CleanupLocationsByHost
- Add SchedulePlanExecutor::SubmitTask for fire-and-forget async work

NOTE: This commit supports V6D heartbeat reporting to KVCM, but does not fully support the V6D bypass-tracker workflow.
Not included:
- StartEvictWriteCache
- V6D metadata lazy deletion
- Select location policy changes
- V6D benchmark tool

### CODE REVIEW

File: kv_cache_manager/data_storage/vineyard_backend.cc [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    KVCM_LOG_INFO("VineyardBackend: node [%s] unregistered from cluster [%s]",
+                  host_ip_port.c_str(),
+                  spec_.cluster_name().c_str());
+    return EC_OK;
+}
+
+void VineyardBackend::OnHeartbeat(const std::string &host_ip_port,
+                                  const std::map<std::string, std::string> &system_status) {
+    std::shared_lock<std::shared_mutex> lock(nodes_mutex_);
+    auto it = nodes_.find(host_ip_port);
```
[the remark this answers is not in the document]
Comment (lucky-zzz) [reply kv_cache_manager/data_storage/vineyard_backend.cc:146]: a given host_ip_port sends heartbeats sequentially, so no concurrent writer for the same NodeInfo.

File: kv_cache_manager/manager/cache_manager.cc
Comment (xiqi-lq) [inline kv_cache_manager/manager/cache_manager.cc:1]: detached std::thread 捕获 this → use-after-free 风险

  File:
  - /tmp/tair-kvcache/kv_cache_manager/manager/cache_manager.cc:1205-1208 (SetCleanupCallback fallback)
  - /tmp/tair-kvcache/kv_cache_manager/manager/cache_manager.cc:1492-1495 (HOST_DOWN fallback)

  两处一模一样的 fallback：

  // L1205 (SetCleanupCallback 的回调内)
  std::thread([this, instance_id, down_host] {
      this->CleanupHostLocations(instance_id, down_host);
  }).detach();

  // L1492 (HOST_DOWN 事件处理)
  std::thread([this, instance_id, host_ip_port] {
      this->CleanupHostLocations(instance_id, host_ip_port);
  }).detach();

  问题在哪：

  detach() 的线程在后台捕获了 this（CacheManager 指针）继续跑。但 CacheManager 一旦析构，这根线程就在访问"野指针"了：

  时间线：
  T+0    : 收到 HOST_DOWN，启动 detached 线程
  T+1ms  : 进程开始 shutdown，CacheManager::~CacheManager() 被调
  T+2ms  : meta_searcher_manager_ / registry_manager_ 被释放
  T+3ms  : detached 线程刚跑到 meta_searcher_manager_->GetMetaSearcher(...)
           → use-after-free，段错误（或者更糟，silently corrupt）

  正路 schedule_plan_executor_->SubmitTask 没这个问题——因为业务线程池的生命周期通常被 CacheManager 持有（用 shared_ptr），析构时会等队列里的任务跑完再释放。

  兜底路径绕开了这个生命周期管理。

File: kv_cache_manager/protocol/protobuf/meta_service.proto [hunk 2]
```diff
@@ -42,6 +42,68 @@ enum StorageType {
...
+
+message NodeRegisterParams {
+    repeated string mediums = 1; // mediums supported by the node, e.g. ["mem","disk"]
+}
+
+message BlockAddParams {
+    string block_key = 1; // int64 as string
+    string uri = 2;       // deprecated, use specs instead
+    string medium = 3;    // medium tag, e.g. "mem"/"disk"/"ssd"/"hbm"; used to build location_id
+    repeated LocationSpec specs = 4; // each entry provides (name, uri) for one spec in the location
```
Comment (xiqi-lq) [inline kv_cache_manager/protocol/protobuf/meta_service.proto]: 和方案设计文档对齐

File: kv_cache_manager/data_storage/vineyard_backend.cc [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+std::vector<bool> VineyardBackend::MightExist(const std::vector<DataStorageUri> &storage_uris) {
+    std::vector<bool> result;
+    result.reserve(storage_uris.size());
+    for (const auto &uri : storage_uris) {
+        if (uri.Valid() && !uri.GetHostName().empty()) {
+            std::string host_ip_port = uri.GetHostName();
+            if (uri.GetPort() > 0) {
+                host_ip_port += ":" + std::to_string(uri.GetPort());
+            }
+            result.push_back(IsNodeRegistered(host_ip_port));
```
Comment (xiqi-lq) [inline kv_cache_manager/data_storage/vineyard_backend.cc:292]: 只看 nodes_ 里有没有，不看 available 字段！
后果：心跳超时 30s → 节点被标 available=false，但 IsNodeRegistered 仍返回 true，KVCM 把 Location 照常返出。要等 5min cleanup_grace 后 UnregisterNode才会被过滤。这 4.5 分钟窗口里，V6D 客户端会拉到已不可达节点的 Location，连接失败。

File: kv_cache_manager/manager/BUILD
Comment (xiqi-lq) [inline kv_cache_manager/manager/BUILD:1]: // cache_manager.cc:1485-1496 
 cleanup 是异步的，UnregisterNode 是同步的。Race 场景：

 T+0    : HOST_DOWN 收到 → SetNodeUnavailable + 排 cleanup 任务 + UnregisterNode
  T+1ms  : V6D 节点重启完成，发新的 EVENT_NODE_REGISTER + EVENT_BLOCK_ADD
  T+2ms  : NODE_REGISTER 成功（nodes_ 里已无旧条目，全新插入）
  T+3ms  : BLOCK_ADD 成功，写入新 Location（block_key=K1，loc_id 含 host_suffix）
  T+5s   : 老的 CleanupHostLocations 才开始 scan
  T+10s  : scan 到 block_key=K1 → 看到 loc_id 后缀匹配 → 误删新写入的 Location

  LivenessCheckerLoop 里也是同样模式（vineyard_backend.cc:248）：先 cleanup_callback，紧接 UnregisterNode。

File: kv_cache_manager/data_storage/vineyard_backend.cc [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            {
+                std::lock_guard<std::mutex> lock(cleanup_cb_mutex_);
+                cb_copy = cleanup_callback_;
+            }
+            for (const auto &host : to_cleanup) {
+                KVCM_LOG_WARN("VineyardBackend: node [%s] passed cleanup_grace_ms, triggering cleanup", host.c_str());
+                if (cb_copy) {
+                    cb_copy(host);
+                }
+                UnregisterNode(host);
```
[the remark this answers is not in the document]
Comment (xiqi-lq) [reply kv_cache_manager/data_storage/vineyard_backend.cc]: 同cache_manager.cc:1485-1496评论，cleanup 是异步的，UnregisterNode 是同步的，会存在race

File: kv_cache_manager/manager/meta_searcher.cc [hunk 5]
```diff
@@ -776,4 +828,144 @@ ErrorCode MetaSearcher::BatchDeleteLocation(RequestContext *request_context,
...
+    }
+
+    if (result.ec != ErrorCode::EC_OK) {
+        KVCM_LOG_WARN("meta_indexer_->ReadModifyWriteLocation failed, ec: %d", result.ec);
+    }
+    return result.ec;
+}
+
+ErrorCode MetaSearcher::CleanupLocationsByHost(RequestContext *request_context,
+                                               const std::string &host_suffix,
```
Comment (xiqi-lq) [inline kv_cache_manager/manager/meta_searcher.cc:924]: CleanupLocationsByHost 不管中间出多少错，最后都 return EC_OK，调用方拿不到任何 partial failure 信号。

File: kv_cache_manager/data_storage/vineyard_backend.cc [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+    }
+    {
+        std::lock_guard<std::mutex> lock(cleanup_cb_mutex_);
+        cleanup_callback_ = nullptr;
+    }
+    KVCM_LOG_INFO("VineyardBackend closed, cluster: [%s]", spec_.cluster_name().c_str());
+    return EC_OK;
+}
+
+void VineyardBackend::SetCleanupCallback(CleanupCallback cb) {
```
Comment (xiqi-lq) [inline kv_cache_manager/data_storage/vineyard_backend.cc:83]: SetCleanupCallback 在每次 ReportEvent 都执行一次，每次 RPC 都 = 一次 mutex lock + 一次 std::function 拷贝 + 一次旧 function 析构。V6D 心跳每 10s 一次，单 instance上百节点时每秒就是几十次重复设置同一个回调。
  建议修法：
  - 在 VineyardBackend 加 std::atomic<bool> cleanup_cb_set_{false}，CacheManager 端先 CAS 检查再 set
  - 或者把注入时机搬到 DataStorageManager 注册 backend 之后

File: kv_cache_manager/data_storage/vineyard_backend.cc (hunk 3 above)
Comment (lucky-zzz) [reply kv_cache_manager/data_storage/vineyard_backend.cc:292]: mightexists函数如果返回false，上层会异步删除该节点的所有元数据，所以在[heartbeat_timeout_ms, cleanup_grace_ms]这段时间窗口的节点目前返回true，不会被过滤也不会被删除。（有部分代码逻辑还没有合入这个pr：在 [heartbeat_timeout_ms, cleanup_grace_ms] 这段"不可用但未死亡"窗口内，应保证 V6D 节点的 Location查询被过滤，但是不做任何删除 ）目前v6d代码中元信息没有下沉到kvcm，这里没有什么影响。

File: kv_cache_manager/manager/cache_manager.cc
Comment (lucky-zzz) [reply kv_cache_manager/manager/cache_manager.cc:1]: 两处 std::thread(...).detach() fallback 移除，改为 assert(schedule_plan_executor_) + 直接走 SubmitTask

File: kv_cache_manager/protocol/protobuf/meta_service.proto (hunk 2 above)
Comment (lucky-zzz) [reply kv_cache_manager/protocol/protobuf/meta_service.proto]: aone已修改

File: kv_cache_manager/manager/meta_searcher.cc (hunk 5 above)
Comment (lucky-zzz) [reply kv_cache_manager/manager/meta_searcher.cc:924]: 已修复

File: kv_cache_manager/manager/BUILD
Comment (lucky-zzz) [reply kv_cache_manager/manager/BUILD:1]: 已修复，思路：
1、在 VineyardBackend 中维护一个 node_generation_ map（跨 UnregisterNode/RegisterNode 生命周期持久化）
2、每次 RegisterNode 时递增 generation
3、触发 cleanup 时捕获当前 generation
4、CleanupHostLocations 检查：如果 generation 已变（说明节点重新注册了），跳过清理

File: kv_cache_manager/data_storage/vineyard_backend.cc (hunk 6 above)
Comment (lucky-zzz) [reply kv_cache_manager/data_storage/vineyard_backend.cc:83]: 按照第一种方法已经修改完

File: kv_cache_manager/data_storage/vineyard_backend.cc [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+                // Re-read with fresh timestamp to avoid racing with OnHeartbeat
+                // which may have updated last_heartbeat_ms after our first read.
+                last_hb = info.last_heartbeat_ms.load(std::memory_order_acquire);
+                int64_t fresh_now = NowMillis();
+                if (fresh_now - last_hb <= heartbeat_timeout_ms_) {
+                    continue;
+                }
+
+                bool prev = info.available.exchange(false, std::memory_order_relaxed);
```
[the remark this answers is not in the document]
Comment (lucky-zzz) [reply kv_cache_manager/data_storage/vineyard_backend.cc:226]: 1、The window is a few nanoseconds wide — requires OnHeartbeat to complete an atomic store in exactly that gap
2、Self-healing: even if triggered, the next OnHeartbeat (≤10s) unconditionally does available.exchange(true) + clears unavailable_since_ms, restoring the node
3、No data loss — only a brief routing suboptimality for at most one heartbeat interval

File: kv_cache_manager/data_storage/vineyard_backend.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucky-zzz) [reply kv_cache_manager/data_storage/vineyard_backend.cc]: 已修

File: kv_cache_manager/service/util/manager_message_proto_util.cc [hunk 8]
```diff
@@ -130,6 +130,13 @@ void ProtoConvert::StorageFromProto(const proto::admin::StorageConfig *proto_sto
         storage_config.set_type(DataStorageType::DATA_STORAGE_TYPE_DUMMY);
         break;
     }
+    case proto::admin::StorageConfig::kVineyard: {
+        VineyardStorageSpec spec;
+        spec.set_cluster_name(proto_storage_config->vineyard().cluster_name());
+        storage_config.set_storage_spec(std::make_shared<VineyardStorageSpec>(spec));
+        storage_config.set_type(DataStorageType::DATA_STORAGE_TYPE_VINEYARD);
```
[the remark this answers is not in the document]
Comment (lucky-zzz) [reply kv_cache_manager/service/util/manager_message_proto_util.cc:141]: 已修复

File: kv_cache_manager/data_storage/vineyard_backend.h [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "kv_cache_manager/data_storage/data_storage_backend.h"
+#include "kv_cache_manager/data_storage/storage_config.h"
+
+namespace kv_cache_manager {
+
+// V6D cluster storage backend. V6D nodes drive liveness via EVENT_HEARTBEAT.
+// LivenessCheckerLoop runs a three-stage decision per node:
+//   healthy -> unavailable (heartbeat timeout) -> dead (cleanup grace exceeded).
+// Does NOT manage allocation (Create/Delete return EC_UNIMPLEMENTED).
```
Comment (xiqi-lq) [inline kv_cache_manager/data_storage/vineyard_backend.h]: 注释还是有点多啊，其他Backend都没啥注释。尽量做到，代码即注释
